opentracing-api/src/main/java/io/opentracing/ActiveSpanHolder.java

```diff
@@ -0,0 +1,174 @@
+     * {@link Continuation#deactivate()}, but rather use {@link ActiveSpanHolder}-aware Runnables/Callables/Executors.
+     * Those higher-level primitives need not be defined within the OpenTracing core API, and so they are not.
+     *
+     * <p>
+     * NOTE: {@link Continuation} extends {@link java.io.Closeable} rather than AutoCloseable in order to keep support
+     * for JDK1.6.
+     *
+     * @see ActiveSpanHolder#capture(Span)
+     */
+    public static abstract class Continuation implements Closeable {
+        private final AtomicInteger refCount;
+
+        protected Continuation(AtomicInteger refCount) {
+            this.refCount = refCount;
+        }
+
+        /**
+         * Make the Span (and other execution context) encapsulated by this Continuation active and return it.
+         *
+         * <p>
+         * NOTE: It is an error to call activate() more than once on a single Continuation instance.
+         *
+         * @see ActiveSpanHolder#capture(Span)
+         * @return the newly-activated Span
+         */
+        public abstract void activate();
+
+        /**
+         * @return the {@link Span} associated with this {@link Continuation}, or null if there is no such {@link Span}.
+         */
+        public abstract Span span();
+
+        /**
+         * Mark the end of this active period for the Span previously returned by activate().
+         *
+         * <p>
+         * NOTE: It is an error to call deactivate() more than once on a single Continuation instance.
+         *
+         * @see Closeable#close()
+         */
+        public final void deactivate() {
+            doDeactivate();
+            decRef();
+        }
+
+        /**
+         * Per the {@link java.io.Closeable} API.
+         */
+        @Override
+        public final void close() {
+            this.deactivate();
+        }
+
+        /**
+         * Capture and take a reference to the {@link Span} associated with this {@link Continuation} and any 3rd-party
+         * execution context of interest.
+         *
+         * @return a new {@link Continuation} to {@link Continuation#activate()} at the appropriate time.
+         */
+        public final Continuation capture() {
+            refCount.incrementAndGet();
+            return holder().doCapture(span(), refCount);
+        }
+
+        /**
+         * Implementations must clean up any state (including thread-locals, etc) associated with the previosly active
+         * {@link Span}.
+         */
+        protected abstract void doDeactivate();
+
+        /**
+         * Return the {@link ActiveSpanHolder} associated wih this {@link Continuation}.
+         */
+        protected abstract ActiveSpanHolder holder();
+
+        /**
+         * Decrement the {@link Continuation}'s reference count, calling {@link Span#finish()} if no more references
+         * remain.
+         */
+        final void decRef() {
+            if (0 == refCount.decrementAndGet()) {
```

Comment:
incRef happens at capture-time, not activation-time.

Re the race condition: I don't think so, at least in practice, as

1. if deactivate() *hasn't* been called for any Continuation, that means the refcount is > 0
2. any capture()ing must happen before deactivate() *is* called
3. thus, the capture/incRef will always happen before the deactivate/decRef